### PR TITLE
add clickhouse connector

### LINKED ISSUE #4500: Create connector for ClickHouse

### PR DESCRIPTION
**Pull request by wgzhao**
The ClickHouse connector allows querying data stored in
Yandex ClickHouse

The ClickHouse connector is compatible with all ClickHouse versions starting from 20.3.5.21.

to view more details in `presto-docs/clichouse.rst`

Fixes #4500

### CODE REVIEW

File: presto-clickhouse/src/main/java/io/prestosql/plugin/clickhouse/ClickHouseConfig.java [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+package io.prestosql.plugin.clickhouse;
+
+import io.airlift.configuration.Config;
+import io.airlift.units.Duration;
+
+import java.util.concurrent.TimeUnit;
+
+public class ClickHouseConfig
+{
+    private String connectionurl;
```
Comment (Praveen2112) [inline presto-clickhouse/src/main/java/io/prestosql/plugin/clickhouse/ClickHouseConfig.java:23]: I guess we don't need this variable

File: presto-clickhouse/src/main/java/io/prestosql/plugin/clickhouse/ClickHouseTableProperties.java [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+    public List<PropertyMetadata<?>> getTableProperties()
+    {
+        return tableProperties;
+    }
+
+    public static Optional<String> getEngine(Map<String, Object> tableProperties)
+    {
+        requireNonNull(tableProperties);
+
+        String value = (String) tableProperties.get(DEFAULT_TABLE_ENGINE);
```
Comment (Praveen2112) [inline presto-clickhouse/src/main/java/io/prestosql/plugin/clickhouse/ClickHouseTableProperties.java:60]: Can we use `Optional#ofNullable`

File: presto-clickhouse/src/main/java/io/prestosql/plugin/clickhouse/ClickHouseTableProperties.java [hunk 3]
```diff
@@ -0,0 +1,66 @@
...
+import java.util.Map;
+import java.util.Optional;
+
+import static io.prestosql.spi.session.PropertyMetadata.stringProperty;
+import static java.util.Objects.requireNonNull;
+
+/**
+ * Class contains all table properties for the Phoenix connector. Used when creating a table:
+ * <p>
+ * <pre>CREATE TABLE foo (a VARCHAR with (primary_key = true), b INT) WITH (SALT_BUCKETS=10, VERSIONS=5, COMPRESSION='lz');</pre>
```
Comment (Praveen2112) [inline presto-clickhouse/src/main/java/io/prestosql/plugin/clickhouse/ClickHouseTableProperties.java:31]: We can have a similar comment for Clickhouse connector

File: presto-clickhouse/src/main/java/io/prestosql/plugin/clickhouse/ClickHouseQueryBuilder.java [hunk 4]
```diff
@@ -0,0 +1,334 @@
...
+import static com.google.common.collect.Iterables.getOnlyElement;
+import static io.prestosql.spi.type.DateTimeEncoding.unpackMillisUtc;
+import static java.lang.Float.intBitsToFloat;
+import static java.util.Collections.nCopies;
+import static java.util.Objects.requireNonNull;
+import static java.util.concurrent.TimeUnit.DAYS;
+import static java.util.stream.Collectors.joining;
+import static org.joda.time.DateTimeZone.UTC;
+
+public class ClickHouseQueryBuilder
```
Comment (Praveen2112) [inline presto-clickhouse/src/main/java/io/prestosql/plugin/clickhouse/ClickHouseQueryBuilder.java:67]: Any reason for extending the `QueryBuilder` , why can't we use the base implementation

File: presto-clickhouse/src/main/java/io/prestosql/plugin/clickhouse/ClickHouseMetadata.java [hunk 5]
```diff
@@ -0,0 +1,234 @@
...
+import java.util.Map;
+import java.util.Optional;
+
+import static com.google.common.collect.ImmutableList.toImmutableList;
+import static java.lang.String.format;
+import static java.lang.String.join;
+import static java.util.Objects.requireNonNull;
+
+public class ClickHouseMetadata
+        extends JdbcMetadata
```
Comment (Praveen2112) [inline presto-clickhouse/src/main/java/io/prestosql/plugin/clickhouse/ClickHouseMetadata.java:54]: Why can't we use JdbcMetadata itself ?

Comment (wgzhao) [reply presto-clickhouse/src/main/java/io/prestosql/plugin/clickhouse/ClickHouseMetadata.java:54]: I need extend `JdbcMetadata` to implement itself `createTable` function .
I want append some table properties like `engine = log` to normal create table sql.
Is there exists simple way to resolve above ?

Comment (Praveen2112) [reply presto-clickhouse/src/main/java/io/prestosql/plugin/clickhouse/ClickHouseMetadata.java:54]: We can override the `BaseJdbcClient#createTable` where we can fetch table properties from `ConnectorTableMetadata`

Comment (wgzhao) [reply presto-clickhouse/src/main/java/io/prestosql/plugin/clickhouse/ClickHouseMetadata.java:54]: thanks, I will try it

Comment (wgzhao) [reply presto-clickhouse/src/main/java/io/prestosql/plugin/clickhouse/ClickHouseMetadata.java:54]: > We can override the `BaseJdbcClient#createTable` where we can fetch table properties from `ConnectorTableMetadata`

hi @Praveen2112  

I try to define table properties in `ClickHouseTableProperties.java`  and as override method `createTable` in `ClickHouseClient.java` , I try get table's properties in the function .

when I  try to create table like the following:

```sql
create table p_tbl(id int) with (engine='Log')
```
the following errors occurred:

````
io.prestosql.spi.PrestoException: Catalog 'clickhouse' does not support table property 'engine'
	at io.prestosql.metadata.AbstractPropertyManager.getProperties(AbstractPropertyManager.java:96)
	at io.prestosql.execution.CreateTableTask.internalExecute(CreateTableTask.java:190)
	at io.prestosql.execution.CreateTableTask.execute(CreateTableTask.java:87)
	at io.prestosql.execution.CreateTableTask.execute(CreateTableTask.java:69)
	at io.prestosql.execution.DataDefinitionExecution.start(DataDefinitionExecution.java:167)
	at io.prestosql.execution.SqlQueryManager.createQuery(SqlQueryManager.java:245)
	at io.prestosql.dispatcher.LocalDispatchQuery.lambda$startExecution$7(LocalDispatchQuery.java:132)
	at io.prestosql.$gen.Presto_339_19_ged4dfb3____20200804_145539_2.run(Unknown Source)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1130)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:630)
	at java.base/java.lang.Thread.run(Thread.java:832)
```

I also find in other connector (such as `phoenix-connector`) , it uses `PhoenixMetadata.java` to process table's properties, is this normal way to do it ?

File: presto-docs/src/main/sphinx/connector/clickhouse.rst [hunk 6]
```diff
@@ -0,0 +1,71 @@
...
+
+    DESCRIBE clichouse.web.clicks;
+    SHOW COLUMNS FROM clichouse.web.clicks;
+
+Finally, you can access the ``clicks`` table in the ``web`` database::
+
+    SELECT * FROM clichouse.web.clicks;
+
+If you used a different name for your catalog properties file, use
+that catalog name instead of ``ClickHouse`` in the above examples.
```
Comment (rosewms) [inline presto-docs/src/main/sphinx/connector/clickhouse.rst:57]: Use the note formatting to better emphasize this info.

File: presto-docs/src/main/sphinx/connector/clickhouse.rst [hunk 7]
```diff
@@ -0,0 +1,71 @@
...
+
+ClickHouse Connector Limitations
+--------------------------------
+
+The following SQL statements are not yet supported:
+
```
Comment (rosewms) [inline presto-docs/src/main/sphinx/connector/clickhouse.rst:63]: Replace with "aren't" - "The following SQL statements aren't supported:"

File: presto-docs/src/main/sphinx/connector/clickhouse.rst [hunk 8]
```diff
@@ -0,0 +1,71 @@
...
+====================
+
+The ClickHouse connector allows querying tables in an external
+`Yandex ClickHouse <https://clickhouse.tech/>`_ instance. This can be used to join data between different
+systems like ClickHouse and Hive, or between two different ClickHouse instances.
+
+Configuration
+-------------
+
+To configure the ClickHouse connector, create a catalog properties file
```
Comment (rosewms) [inline presto-docs/src/main/sphinx/connector/clickhouse.rst:12]: Try "To configure the ClickHouse connector, create a catalog properties file ``etc/catalog/clickhouse.properties``. Replace the connection properties as appropriate for your setup:"

**Pull request closed without merging**